assets/backport/github/github.go

```diff
@@ -0,0 +1,398 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package github
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gravitational/trace"
+
+	go_github "github.com/google/go-github/v37/github"
+	"golang.org/x/oauth2"
+)
+
+type Client struct {
+	Client *go_github.Client
+	Config
+}
+
+type Config struct {
+	Token        string
+	Organization string
+	Repository   string
+	Username     string
+}
+
+// New returns a new GitHub client.
+func New(ctx context.Context, c Config) (*Client, error) {
+	err := validateConfig(c)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: c.Token},
+	)
+	return &Client{
+		Client: go_github.NewClient(oauth2.NewClient(ctx, ts)),
+		Config: c,
+	}, nil
+}
+
+// validateConfig validates config.
+func validateConfig(c Config) error {
+	if c.Token == "" {
+		return trace.BadParameter("missing token")
+	}
+	if c.Organization == "" {
+		return trace.BadParameter("missing organization")
+	}
+	if c.Repository == "" {
+		return trace.BadParameter("missing repository")
+	}
+	if c.Username == "" {
+		return trace.BadParameter("missing username")
+	}
+	return nil
+}
+
+// Backport backports changes from backportBranchName to a new branch based
+// off baseBranchName.
+//
+// A new branch is created with the name in the format of
+// auto-backport/[baseBranchName]/[backportBranchName], and
+// cherry-picks commits onto the new branch.
+func (c *Client) Backport(ctx context.Context, baseBranchName, backportBranchName string, commits []*go_github.Commit) (string, error) {
+	newBranchName := fmt.Sprintf("auto-backport/%s/%s", baseBranchName, backportBranchName)
+	// Create a new branch off of the target branch.
+	err := c.createBranchFrom(ctx, baseBranchName, newBranchName)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	fmt.Printf("Created a new branch: %s.\n", newBranchName)
+
+	// Cherry pick commits.
+	err = c.cherryPickCommitsOnBranch(ctx, newBranchName, commits)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	fmt.Printf("Finished cherry-picking %v commits. \n", len(commits))
+	return newBranchName, nil
+}
+
+// cherryPickCommitsOnBranch cherry picks a list of commits on a given branch.
+func (c *Client) cherryPickCommitsOnBranch(ctx context.Context, branchName string, commits []*go_github.Commit) error {
+	branch, err := c.getBranch(ctx, branchName)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	headCommit, err := c.getCommit(ctx, *branch.Commit.SHA)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	for i := 0; i < len(commits); i++ {
+		tree, sha, err := c.cherryPickCommit(ctx, branchName, commits[i], headCommit)
+		if err != nil {
+			defer c.deleteBranch(ctx, branchName)
+			return trace.Wrap(err)
+		}
+		headCommit.SHA = &sha
+		headCommit.Tree = tree
+	}
+	return nil
+}
+
+// cherryPickCommit cherry picks a single commit on a branch.
```

Comment:
I'd add some extra commentary here about what all the hoops it has to jump through to create a cherry-pick, something like:

> Cherry-picking is not a simple operation in git. The actual algorithm is quite complicated:
> 1. Create a _sibling commit_ (i.e. a commit that shares a parent) to the _source commit_ (i.e. the
>  commit we want to cherry-pick onto another branch). This _sibling commit_ has the same file tree as
>  the HEAD of the _destination branch_ (the branch that the cherry-pick will end up on if all
>  goes well).
> 2. Merge the _source commit_ onto the _sibling commit_, which creates a new new _merge 
>    commit_ that contains only the delta between the file tree in _target branch_ and the _source commit_.
> 3. Create the actual _cherry-pick commit_  on the _target branch_ containing the file tree delta generated from  the _merge commit_. 
> 4. The _sibling_ and _merge_ commits can now be discarded.